src/main/java/com/cleanroommc/groovyscript/server/features/textureDecoration/TextureDecoration.java

```diff
@@ -0,0 +1,108 @@
+package com.cleanroommc.groovyscript.server.features.textureDecoration;
+
+import com.cleanroommc.groovyscript.mapper.AbstractObjectMapper;
+import com.cleanroommc.groovyscript.mapper.TooltipEmbedding;
+import net.minecraft.client.renderer.GlStateManager;
+import org.codehaus.groovy.ast.expr.ArgumentListExpression;
+import org.codehaus.groovy.ast.expr.ConstantExpression;
+import org.jetbrains.annotations.NotNull;
+import org.lwjgl.opengl.GL11;
+
+import javax.imageio.ImageIO;
+import java.awt.image.BufferedImage;
+import java.io.File;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.List;
+
+public class TextureDecoration<T> {
+
+    private static <T> T getObjectWithConstArgs(AbstractObjectMapper<T> mapper, ArgumentListExpression args) {
+        var additionalArgs = new Object[args.getExpressions().size() - 1];
+        for (int i = 0; i < additionalArgs.length; i++) {
+            if (args.getExpressions().get(i + 1) instanceof ConstantExpression argExpression)
+                additionalArgs[i] = argExpression.getValue();
+        }
+        return mapper.invoke(true, args.getExpressions().get(0).getText(), additionalArgs);
+    }
+
+    private final String name;
+    private final String uri;
+    private final AbstractObjectMapper<T> mapper;
+    private final T bindable;
+    boolean fileExists;
+    boolean queued;
+
+    public static <T> TextureDecoration<T> create(String name, String uri, AbstractObjectMapper<T> mapper, ArgumentListExpression args) {
+        T bindable = getObjectWithConstArgs(mapper, args);
+        if (bindable == null) return null;
+        return new TextureDecoration<>(name, mapper, bindable, uri);
+    }
+
+    public static <T> TextureDecoration<T> of(TooltipEmbedding<T> embedding) {
+        return new TextureDecoration<>(embedding.getTextureName(), embedding.getMapper(), embedding.getContext(), TextureDecorationProvider.getURIForDecoration(embedding.getTextureName()));
+    }
+
+    private TextureDecoration(String name, AbstractObjectMapper<T> mapper, T bindable, String uri) {
+        this.name = name;
+        this.mapper = mapper;
+        this.bindable = bindable;
+        this.uri = uri;
+        File file = getFile();
+        this.fileExists = !file.getParentFile().mkdirs() && file.exists();
+    }
+
+    public void bindTexture() {
+        this.mapper.bindTexture(this.bindable);
+    }
+
+    public List<String> getTooltip() {
+        List<String> tooltip = this.mapper.getTooltip(this.bindable);
+        return tooltip.isEmpty() ? tooltip : new ArrayList<>(tooltip);
```

Comment:
whats the difference between these? why `new ArrayList<>(tooltip)` if `tooltip` is empty?
```diff
-tooltip.isEmpty() ? tooltip : new ArrayList<>(tooltip)
+tooltip.isEmpty() ? tooltip : new ArrayList<>()
```

Reply:
No. Return it if empty and copy if not.

Reply:
right sorry misread, had that flipped in my head as "if its not empty return it, otherwise return an empty list".